src/lib/protocols/cassandra.c

```diff
@@ -18,125 +21,69 @@
-    case CASSANDRA_V4_RESPONSE:
-      return true;
-  }
-  return false;
+  return (response >= 0x81 && response <= 0x84) ? 1 : -1;
 }
 
-static bool ndpi_check_valid_cassandra_opcode(uint8_t opcode)
+static inline int ndpi_validate_cassandra_request(u_int8_t request)
 {
-  switch (opcode) {
-    case CASSANDRA_ERROR:
-    case CASSANDRA_STARTUP:
-    case CASSANDRA_READY:
-    case CASSANDRA_AUTHENTICATE:
-    case CASSANDRA_OPTIONS:
-    case CASSANDRA_SUPPORTED:
-    case CASSANDRA_QUERY:
-    case CASSANDRA_RESULT:
-    case CASSANDRA_PREPARE:
-    case CASSANDRA_EXECUTE:
-    case CASSANDRA_REGISTER:
-    case CASSANDRA_EVENT:
-    case CASSANDRA_BATCH:
-    case CASSANDRA_AUTH_CHALLENGE:
-    case CASSANDRA_AUTH_RESPONSE:
-    case CASSANDRA_AUTH_SUCCESS:
-      return true;
-  }
-  return false;
+  return (request >= 0x01 && request <= 0x04) ? 1 : -1;
 }
 
-static bool ndpi_check_valid_cassandra_flags(uint8_t flags)
+static void ndpi_int_cassandra_add_connection(struct ndpi_detection_module_struct *ndpi_struct,
+                                              struct ndpi_flow_struct *flow)
 {
-  return (flags & 0xF0) == 0;
+  NDPI_LOG_INFO(ndpi_struct, "found Cassandra CQL\n");
+
+  ndpi_set_detected_protocol(ndpi_struct, flow,
+                             NDPI_PROTOCOL_CASSANDRA, NDPI_PROTOCOL_UNKNOWN,
+                             NDPI_CONFIDENCE_DPI);
 }
 
 static void ndpi_search_cassandra(struct ndpi_detection_module_struct *ndpi_struct,
                                   struct ndpi_flow_struct *flow)
 {
   struct ndpi_packet_struct *packet = &ndpi_struct->packet;
 
-  NDPI_LOG_DBG(ndpi_struct, "search Cassandra\n");
+  NDPI_LOG_DBG(ndpi_struct, "search Cassandra CQL\n");
 
-  if (packet->tcp) {
-    if (packet->payload_packet_len >= CASSANDRA_HEADER_LEN &&
-        ndpi_check_valid_cassandra_version(get_u_int8_t(packet->payload, 0)) &&
-        ndpi_check_valid_cassandra_flags(get_u_int8_t(packet->payload, 1)) &&
-        ndpi_check_valid_cassandra_opcode(get_u_int8_t(packet->payload, 4)) &&
-        ntohl(get_u_int32_t(packet->payload, 5)) <= CASSANDRA_MAX_BODY_SIZE &&
-        ntohl(get_u_int32_t(packet->payload, 5)) >= (uint32_t) (packet->payload_packet_len - CASSANDRA_HEADER_LEN) &&
-        flow->h323_valid_packets == 0 /* To avoid clashing with H323 */ &&
-        flow->socks4_stage == 0 /* To avoid clashing with SOCKS */) {
-      if (flow->packet_counter > 3)
-      {
-        NDPI_LOG_INFO(ndpi_struct, "found Cassandra\n");
-        ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_CASSANDRA, NDPI_PROTOCOL_UNKNOWN, NDPI_CONFIDENCE_DPI);
-      }
-      return;
-    }
+  if (packet->payload_packet_len < 9 ||
+      (!ndpi_validate_cassandra_response(packet->payload[0]) ||
+       !ndpi_validate_cassandra_request(packet->payload[0])))
+  {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
+  
+  if (flow->packet_counter > 12) {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
```

Comment:
```4.1.1. STARTUP: This must be the first message of the connection, except for OPTIONS that can be sent before to find out the options supported by the server.```

Since in the worst case, we are looking for the second packets sent by the client I think that we might be able to have a better/faster check. Something like:
```if(flow->packet_direction_counter[packet->packet_direction] > 2 { /*exclude */```

Reply:
Thanks for the review, it really works better with this check.